libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/ReasoningBasedSimplifier.h>
+
+#include <libyul/optimiser/SSAValueTracker.h>
+#include <libyul/AsmData.h>
+#include <libyul/Utilities.h>
+#include <libyul/Dialect.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <libsmtutil/Z3Interface.h>
+
+#include <libsolutil/Visitor.h>
+#include <libsolutil/CommonData.h>
+
+#include <utility>
+#include <memory>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::smtutil;
+
+void ReasoningBasedSimplifier::run(OptimiserStepContext& _context, Block& _ast)
+{
+	set<YulString> ssaVars = SSAValueTracker::ssaVariables(_ast);
+	ReasoningBasedSimplifier{_context.dialect, ssaVars}(_ast);
+}
+
+void ReasoningBasedSimplifier::operator()(VariableDeclaration& _varDecl)
+{
+	// TODO assign zero if no value.
+	if (_varDecl.variables.size() != 1 || !_varDecl.value)
+		return;
+	YulString varName = _varDecl.variables.front().name;
+	if (!m_ssaVariables.count(varName))
+		return;
+	m_variables.insert({varName, m_solver->newVariable("yul_" + varName.str(), SortProvider::bitVectorSort)});
+	m_solver->addAssertion(m_variables.at(varName) == encodeExpression(*_varDecl.value));
+}
+
+void ReasoningBasedSimplifier::operator()(If& _if)
+{
+	smtutil::Expression condition = encodeExpression(*_if.condition);
+	m_solver->push();
+	m_solver->addAssertion(condition == int2bv(size_t(0)));
+	CheckResult result = m_solver->check({}).first;
+	m_solver->pop();
+	if (result == CheckResult::UNSATISFIABLE)
+		_if.condition = make_unique<yul::Expression>(Literal{locationOf(*_if.condition), LiteralKind::Number, "1"_yulstring, {}});
+
+	m_solver->push();
+	m_solver->addAssertion(condition != int2bv(size_t(0)));
+	CheckResult result2 = m_solver->check({}).first;
+	m_solver->pop();
+	if (result2 == CheckResult::UNSATISFIABLE)
+		// TODO we could actually skip the body is this case
```

Comment:
Shouldn't the `StructuralSimplifier` do that for us?

Also if we rely on that, probably no point supporting switches here, but if we don't probably we should support some potential switch cases, however that feels like a complicated piece of code.

Reply:
Yea makes sense that here we just replace the condition expression by a constant, and then another simplifier does the body removal.

Reply:
However... we're replacing the condition here, so that it's replaced again later. Might as well remove the whole body here.

Reply:
Probably would prefer to have a helper in `OptimizerUtils.cpp` which does that, like `removeNode(Switch)`.